Added support for LIST_REVERSE function

MaDChronex opened the pull request:
Added a function with a list as input. The function returns the reverse of the input. Example follows:
```
SELECT LIST_REVERSE (ARRAY[1, 3, 2, 42, 117, 68, NULL]);
```
```
┌───────────────────────────────────────────────────┐
│ list_reverse((ARRAY[1, 3, 2, 42, 117, 68, NULL])) │
│                      int32[]                      │
├───────────────────────────────────────────────────┤
│ [NULL, 68, 117, 42, 2, 3, 1]                      │
└───────────────────────────────────────────────────┘

```

Commit:
Add support for list_reverse function

Commit:
added tests and removed rows in function

Commit:
fixed unique to uniq issue

Commit:
format-fix list_reverse.cpp

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,117 @@
+		}
+		result_entries[i].offset = offset;
+		result_entries[i].length = 0;
+		if (input_list_data.validity.RowIsValid(input_list_list_index)) {
```
Didn't we already make sure this is true by dealing with NULL earlier?

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,117 @@
+
+	auto &input_list = arguments[0]->return_type;
+	if (input_list.id() == LogicalTypeId::UNKNOWN) {
+		throw ParameterNotResolvedException();
```
Can we add a test that calls this function with a prepared parameter?
```sql
statement ok
prepare q1 as select list_reverse(?);

query I
execute q1([5,4,3]);
----
[3,4,5]
```

Tishj commented on test/sql/types/nested/list/test_list_reverse.test:
```diff
@@ -0,0 +1,45 @@
+# name: test/sql/types/nested/list/test_list_reverse.test
+# description: Test reversing the order of the elements of a list
+# group: [list]
```
Can we add some coverage tests with different vector types?
You can use the `test_vector_types` function to help with this, see `test/sql/types/nested/map/test_map_subscript_vector_types.test` for an example.

Also can you try with a column as input instead of a constant expression?
(using something bigger than 2048 here just to also test how this behaves with lists bigger than STANDARD_VECTOR_SIZE)
```sql
statement ok
create table tbl as select range(5000) as list;

query I nosort reversed_list
select list_sort(range(5000), 'desc');

query I nosort reversed_list
select list_reverse(list) from tbl;
```

taniabogatsch reviewed (changes requested):
Hi @MaDChronex, thanks for your contribution!

First, I have a performance-related question about your approach. Currently, we append each reversed list individually while looping over all rows. I didn't benchmark this, but it might become very slow if we face many lists with only a few elements.

I did not think as much about this function as you did, so please take a critical look at my suggestion. Each list's length and offset do not change during this operation. So you can copy the list entries vector into the result vector. Then, you can use our slice method `Vector::Slice` to slice the child vector into the result vector's child vector. Because you already construct the selection vector for the entire child vector. I believe this allows us to completely omit the for-loop. 

Additionally, it'd be nice if you could add additional tests.
- Create tables with multiple rows, reversing all the rows in a column
- Tests with nested types (`[[2,3], [NULL]]` -> `[[NULL], [2,3]]`
- Optionally, together with lambda functions, and with different data types

douenergy commented:
@MaDChronex  Maybe the test case here will be helpful.

https://github.com/duckdb/duckdb/pull/7081/files 

- list_reverse(NULL)
- ([1,5,3,9]).list_reverse()
- list_reverse( list_reverse([1,5,3,9]) )
- test from [Bigquery](https://cloud.google.com/bigquery/docs/reference/standard-sql/array_functions#array_reverse)

taniabogatsch commented:
I thought a bit more about this. We cannot construct the selection vector for multiple rows as we currently do. E.g., given the following list vector and child vector
```
[0, 1, 2, 3]
[4, 5, 6]
[7, 8]

offset, length: (0, 4), (4, 3), (7, 2), child vector: (0, 1, 2, 3, 4, 5, 6, 7, 8)
reversed child vector (also the selection vector): (3, 2, 1, 0, 6, 5, 4, 8, 7)
```
So we still need to loop over the rows to get the length of each list to correctly construct the selection vector. We can also set the `list_entry_t's` of the result vector while doing that. And then, we slice the child vector with that selection vector after the loop. 🤔

Commit:
Applied feedback (thanks!)
- Reserved result vector beforehand
- removed obsolete code and added D_ASSERT
- Added additional tests

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -40,19 +40,15 @@ static void ListReverseFunction(DataChunk &args, ExpressionState &state, Vector
 	for (idx_t i = 0; i < count; i++) {
 		auto input_list_list_index = input_list_data.sel->get_index(i);
 
-		if (!input_list_data.validity.RowIsValid(input_list_list_index)) {
```
That might be a little too eager of a change, no?
The comment was about removing the check after this, *because* this check ensures that it's always true

Unless this is not needed because of something else that I'm not following?

MaDChronex replied:
Hello @Tishj,
Thanks again for the fast feedback. I may have cleaned up a bit too much. I wasn't running into any issues for some reason. I will add another test for mistakes in the validity map. Thank you again!

Mytherin commented on test/sql/types/nested/list/test_list_reverse.test:
```diff
@@ -0,0 +1,67 @@
+[2, NULL, 42, 1]
+
+query I
+SELECT LIST_REVERSE (LIST_REVERSE(ARRAY[1, 3, 3, 42, 117, 69, NULL]))
```
Could we add some tests with data in a table? e.g.:

```sql
CREATE TABLE lists(l INT[]);
INSERT INTO lists VALUES ([1, 2, 3]), (NULL), ([]), ([NULL, NULL, 5, 6, NULL]);
SELECT list_reverse(l) FROM lists;
```

And tests with `test_vector_types` and `test_all_types`, e.g.::

```sql
select list_reverse(columns(['int_array', 'varchar_array', 'nested_int_array', 'array_of_structs', 'timestamp_array', 'double_array', 'date_array', 'timestamptz_array'])) from test_all_types();
select list_reverse(test_vector) from test_vector_types(null::int[], false);
select list_reverse(test_vector) from test_vector_types(null::int[], true);
```

Commit:
fixed issue with mutiple rows and added test cases
issues with list_reverse(test_vector) from test_vector_types still prevails

MaDChronex committed:
Merge branch 'master' into list_reverse

MaDChronex committed:
Merge branch 'duckdb:master' into list_reverse

Tishj commented:
To continue on Tanias suggestion:

Since the elements of the Vector are the same, we can use a DictionaryVector for this.
What a dictionary vector does is essentially map from one index to another.

Let's take this example:
```
[2, 3, 5]
[1, 2]
NULL
[],
[],
[2,3,4,5]
```
The child Vector of this contains:
```
data : [2,3,5,1,2,2,3,4,5]
index: [0,1,2,3,4,5,6,7,8]
```

When we call this function, we end up with this order:
```
data : [5,3,2,2,1,5,4,3,2]
index: [2,1,0,4,3,8,7,6,5]
```

The elements are the same, only the indices change, which is exactly what a SelectionVector is for.
Once you have made the SelectionVector that represents this, you should be able to just apply it to the child vector.
```c++
// populate the sel ...

result.Reference(input_list);
auto &child = ListVector::GetEntry(result);
child.Slice(sel, ListVector::GetListSize(input_list);
```

MaDChronex committed:
Merge branch 'duckdb:master' into list_reverse

MaDChronex committed:
Merge branch 'duckdb:master' into list_reverse

Commit:
Discarded append and used slice

MaDChronex commented:
It is working fine and more compact using the slice method. Thank you for helping me on that front! however I still have issues specifically with: 
```
select list_reverse(test_vector) from test_vector_types(null::int[], false);
select list_reverse(test_vector) from test_vector_types(null::int[], true);
``` 
The problem I have is that usually the child vector of multiple input rows is one single child vector. Tishj’s example:
```
[2, 3, 5]
[1, 2]
NULL
[],
[],
[2,3,4,5]
```
The single child vector looks like:
```[2,3,5,1,2,2,3,4,5]```
That is not the case with test_vector_types, where I count multiple child vectors, which results in issues building the reversed selection vector, as it relies on the overall size of the child vector.
I do not fully understand why test_vector_types behaves like this. Do I need to have list_indices similar to list_sort?
Thank you for all the suggestions so far!

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,124 @@
+
+	// get the child vector
+	auto input_list_list_size = ListVector::GetListSize(input_list);
+	auto &input_list_child = ListVector::GetEntry(input_list);
```
we're not interested in the list_child, so this can be left out

MaDChronex replied:
We need input_list_child as we slice it later. Is there a way where we do not slice the child vector?

Tishj replied:
By slicing the result child instead

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,124 @@
+
+	UnifiedVectorFormat input_list_data;
+	input_list.ToUnifiedFormat(count, input_list_data);
+	auto input_list_entries = (list_entry_t *)input_list_data.data;
```
please use `UnifiedVectorFormat::GetData<list_entry_t>(input_list_data);` to get the list entries
We're in the process of banning all c-style casts :)

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,124 @@
+
+	result.SetVectorType(VectorType::FLAT_VECTOR);
+	auto result_entries = FlatVector::GetData<list_entry_t>(result);
+	auto &result_validity = FlatVector::Validity(result);
```
Nor this, as this is also inherited through `Reference`

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,124 @@
+		auto input_list_list_index = input_list_data.sel->get_index(i);
+
+		if (!input_list_data.validity.RowIsValid(input_list_list_index)) {
+			result_validity.SetInvalid(i);
```
because we don't set the result validity manually, this is also not needed

MaDChronex replied:
It appears that we need to do that to handle cases like ```select list_reverse(columns(['int_array', 'varchar_array', 'nested_int_array', 'array_of_structs', 'timestamp_array', 'double_array', 'date_array', 'timestamptz_array'])) from test_all_types();
```

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,124 @@
+		// set reverse selection vector indices
+		// set index of selection vector in a way, that only the entries of the current row are reversed at once
+		for (idx_t j = 0; j < input_list_entry.length; j++) {
+			rev_sel.set_index(j + offset_per_new_row, input_list_entry.length - j - 1 + offset_per_new_row);
```
the `offset_per_new_row` is the reason you were having difficulties with VectorType's other than FLAT, because ConstantVectors only have 1 physical row, but `count` logical rows, this would overflow the selection vector.

If you instead use `input_list_entry.offset` for this, that issue is solved

MaDChronex replied:
Thats it! That helps a lot. Thank you so much!!!

Tishj commented on test/sql/types/nested/list/test_list_reverse.test:
```diff
@@ -0,0 +1,109 @@
+NULL	NULL	NULL	NULL	NULL	NULL	NULL	NULL
+
+#TODO: make this test not fail.
+#select list_reverse(test_vector) from test_vector_types(null::int[], false);
```
These create ConstantVector and also DictionaryVector, which was breaking your selection vector reverse logic, these should pass now.

`test_vector_types` outputs rows based on the STANDARD_VECTOR_SIZE, and we have tests for different values for the standard vector size, which would break a test that did explicit result checking for this.

Instead you probably want to use this:
```sql
query I nosort q1
select true from test_vector_types(null::int[], false);
----

query I nosort q1
select list_sort((list), 'desc') == list from (select list_reverse(test_vector) as list from test_vector_types(null::int[], false));
----

query I nosort q1
select list_sort((list), 'desc') == list from (select list_reverse(test_vector) as list from test_vector_types(null::int[], true));
----
```

using the labels (`q1` in this case) we assert that the results of the two queries are identical, without having to set an explicit expected result.

Tishj reviewed (changes requested):
Thanks for the updates, it looks good!

Just a couple of things that could be removed and a couple notes on bugs in your implementation and how to address them.

MaDChronex committed:
Merge branch 'duckdb:master' into list_reverse

Commit:
added tests, solved the issue created by offset_per_new_row. Issues caused by comparison with list_sort prevail.

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,115 @@
+
+		D_ASSERT(input_list_data.validity.RowIsValid(input_list_list_index));
+		const auto &input_list_entry = input_list_entries[input_list_list_index];
+		result_entries[i].length += input_list_entry.length;
```
These entries are likely uninitialized, and as stated before, this step is just completely not necessary as both the length and the offsets for each list will be unchanged from the input

Tishj commented on src/function/scalar/list/list_reverse.cpp:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	result.Reference(input_list);
+	input_list_child.Slice(rev_sel, ListVector::GetListSize(input_list));
```
I think your issue with the comparison to `list_sort(..)` result stems from changing the input list here.
If you instead apply the Slice to the `ListVector::GetEntry(result)` that should fix it

MaDChronex replied:
Slicing the result entries did not fix it entirely, but the issue was indeed that the input is changed here. So I took a page out of list_sort and used:
‘’’
VectorOperations::Copy(input_list, result, count, 0, 0);
‘’’
Using that and working entirely on result and not the input list solved the issue!
Thanks a lot for the help!

Commit:
fixed the comparison issue

Mytherin commented:
This PR has been superseded by #8477
